File: server/src/derive.cpp
#include "derive.hpp"
#include "hmac_verify.hpp"

#include <stdexcept>

namespace knockd {

std::vector<uint16_t> derive_ports(const std::vector<uint8_t> &secret,
                                    uint64_t time_counter,
                                    int n,
                                    uint16_t port_low,
                                    uint16_t port_high,
                                    uint8_t channel_id) {
    if (n < 1 || n > 16) {
        throw std::invalid_argument("n must be between 1 and 16");
    }
    if (port_high <= port_low) {
        throw std::invalid_argument("port_high must be greater than port_low");
    }
    if (channel_id >= kNumChannels) {
        throw std::invalid_argument("channel_id must be less than kNumChannels");
    }

    std::vector<uint8_t> msg(9);
    for (int i = 0; i < 8; ++i) {
        msg[7 - i] = static_cast<uint8_t>((time_counter >> (8 * i)) & 0xFF);
    }
    msg[8] = channel_id;

    std::array<uint8_t, 32> digest = hmac_sha256(secret, msg);

    uint32_t span = static_cast<uint32_t>(port_high) - static_cast<uint32_t>(port_low) + 1;
    std::vector<uint16_t> ports;
    ports.reserve(static_cast<size_t>(n));
    for (int i = 0; i < n; ++i) {
        uint16_t raw16 = static_cast<uint16_t>((digest[2 * i] << 8) | digest[2 * i + 1]);
        ports.push_back(static_cast<uint16_t>(port_low + (raw16 % span)));
    }
    return ports;
}

} // namespace knockd

File: client/knockc/derive.py
"""Knock sequence derivation. Mirrors spec/derivation.md and server/src/derive.cpp
byte-for-byte -- see spec/testvectors.json for the cross-language parity tests.

Protocol v2: the HMAC message is time_counter (8 bytes, big-endian)
concatenated with channel_id (1 byte). channel_id 0 opens its configured
target port(s) (the entire v1 behavior, now possibly more than one port);
1-10 each run a server-configured action -- a command, or opening that
channel's own port(s).
"""
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.hmac import HMAC

NUM_CHANNELS = 11  # valid channel_id values: 0..10


def time_counter(unix_time: float, step_seconds: int) -> int:
    return int(unix_time // step_seconds)


def compute_digest(secret: bytes, counter: int, channel_id: int = 0) -> bytes:
    if not (0 <= channel_id < NUM_CHANNELS):
        raise ValueError(f"channel_id must be in 0..{NUM_CHANNELS - 1}")
    msg = counter.to_bytes(8, "big") + bytes([channel_id])
    h = HMAC(secret, hashes.SHA256())
    h.update(msg)
    return h.finalize()


def ports_from_digest(digest: bytes, n: int, port_low: int, port_high: int) -> list[int]:
    if not (1 <= n <= 16):
        raise ValueError("n must be between 1 and 16 (SHA-256 digest is 16 two-byte chunks)")
    if port_high <= port_low:
        raise ValueError("port_high must be greater than port_low")

    span = port_high - port_low + 1
    ports = []
    for i in range(n):
        raw16 = int.from_bytes(digest[2 * i:2 * i + 2], "big")
        ports.append(port_low + (raw16 % span))
    return ports


def derive_ports(secret: bytes, counter: int, n: int, port_low: int, port_high: int,
                  channel_id: int = 0) -> list[int]:
    digest = compute_digest(secret, counter, channel_id)
    return ports_from_digest(digest, n, port_low, port_high)
